### ISSUE TITLE
Endless loop in deque::shrink_to_fit()

### ISSUE THREAD

**Issue by CopaDataDevelopment** [microsoft/STL#4954]
# Describe the bug
Commit c40a251887853dee30c406bbb94522b35195d912 (fixing issue #4091, merged into master in issue #4072) introduced an endless loop into deque::shrink_to_fit(). In specific constellations in regards to _First_used_block_idx, _First_unused_block_idx and _Mask the first loop for deallocating unused blocks runs infinetely.

# Command-line test case
```
#define _CRT_SECURE_NO_WARNINGS

#include <windows.h>
#include <math.h>
#include <deque>

int main()
{
  std::deque<int> qu;

  long it = 0;
  while (1)
  {
    size_t numAlloc = rand() + 1;
    for (size_t i = 0; i < numAlloc; i++)
    {
      qu.push_back(0);
    }

    size_t numDealloc = rand() + 1;
    if (it % 100 == 0 || numDealloc > qu.size())
    {
      numDealloc = qu.size();
    }
    for (size_t i = 0; i < numDealloc; i++)
    {
      qu.pop_front();
    }
    qu.shrink_to_fit();

    printf("iteration %d: %lld\n", ++it, qu.size());
  }

  return 0;
}
```
After about 40 iterations deque::shrink_to_fit get's stuck.

# Expected behavior
Termination condition for loop is correct so deque::shrink_to_fit() terminates.

# STL version
* Visual Studio version
    ```
    Microsoft Visual Studio Enterprise 2022 
    Version 17.10
    ```
* Visual Studio version
    ```
    Microsoft Visual Studio Enterprise 2022 
    Version 17.11.1
    ```

**Comment by frederick-vs-ja** [comment]
`_First_used_block_idx` was incorrectly unmasked. In the problematic case, `_First_used_block_idx` became larger than `_Mask`, so the loop never ended. I'm fixing this.

**Comment by CDAlexanderW** [comment]
Can we have this fix please backported to 17.11.x?

**Comment by StephanTLavavej** [comment]
Given the timing of when this was reported, I don't believe that 17.11.x will be feasible (as it's odd-numbered and not long-term support). We can likely get it into 17.12 before GA. It *may* be worth backporting to 17.10, the original release where it regressed, although that's triple the work.

**Comment by dmayola** [comment]
Hello, I encountered this issue in 17.11.5, I haven't gotten into all the details or a potential fix, but we've seen some random freezes in our application due to an endless loop in:

```C++
        // deallocate unused blocks, traversing over the circular buffer until the first used block index
        for (auto _Block_idx = _First_unused_block_idx; _Block_idx != _First_used_block_idx;
             _Block_idx      = static_cast<size_type>((_Block_idx + 1) & _Mask)) {
            auto& _Block_ptr = _Map()[static_cast<_Map_difference_type>(_Block_idx)];
            if (_Block_ptr != nullptr) {
                _Getal().deallocate(_Block_ptr, _Block_size);
                _Block_ptr = nullptr;
            }
        }
```

**Comment by CDAlexanderW** [comment]
Can you confirm this fix only landed in 17.13. preview 1 and not in 17.12.1?
And also that it won't show up in another 17.12.x update?

Just curious and asking because last month I saw the following in the maintainers priority (https://github.com/microsoft/STL/issues/4700) of @CaseyCarter. Nevertheless I didn't find any additional transparent information where this landed (except the STL Changelog https://github.com/microsoft/STL/wiki/Changelog, but I don't expect any fixes to older versions documented there).

### :warning: **High priority**
* Backport #4955 to 17.12 and 17.10

**Comment by StephanTLavavej** [comment]
Correct, we were thinking about backporting the fix, but we were too busy and dropped those plans. (This is a relatively obscure function, with a simple workaround, and it's a problem that will "solve itself" when 17.13 ships for production.)
